### PR TITLE
Smaato: DOOH support

### PR DESCRIPTION
**Pull request by Enigo**
Adding DOOH support for Smaato adapater

### CODE REVIEW

File: adapters/smaato/smaato.go [hunk 1]
```diff
@@ -448,8 +456,12 @@ func setPublisherId(request *openrtb2.BidRequest, imp *openrtb2.Imp) error {
 		// App is already a copy
 		request.App.Publisher = &openrtb2.Publisher{ID: publisherID}
 		return nil
+	} else if request.DOOH != nil {
+		// App is already a copy
```
Comment (el-chuck) [inline adapters/smaato/smaato.go:460]: "App" to "DOOH"

File: adapters/smaato/banner.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+func extractAdmBanner(adMarkup string, curls []string) string {
+	var clickEvent string
+	if len(curls) > 0 {
+		var clicks strings.Builder
+		for _, clicktracker := range curls {
```
Comment (SyntaxNode) [inline adapters/smaato/banner.go:13]: How many curls do you expect the adm to contain?

File: adapters/smaato/banner_test.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.testName, func(t *testing.T) {
+			adMarkup := extractAdmBanner(tt.adMarkup, tt.curls)
+
+			assert.Equal(t, tt.expectedAdMarkup, adMarkup)
+		})
+	}
+}
```
Comment (SyntaxNode) [inline adapters/smaato/banner_test.go:42]: This is fine, but doesn't count towards the test coverage for our review purposes. Please ensure these cases are covered with the json test framework.

File: adapters/smaato/smaato.go [hunk 4]
```diff
@@ -57,7 +57,8 @@ type bidRequestExt struct {
 
 // bidExt defines Bid.Ext object for Smaato
 type bidExt struct {
-	Duration int `json:"duration"`
+	Duration int      `json:"duration"`
+	Curls    []string `json:"curls"`
```
Comment (SyntaxNode) [inline adapters/smaato/smaato.go:61]: What is a Curl?

File: adapters/smaato/smaato.go [hunk 5]
```diff
@@ -260,20 +266,12 @@ func (adapter *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.Req
...
-	} else if strings.HasPrefix(adMarkup, `{"richmedia":`) {
-		return smtAdTypeRichmedia, nil
-	} else if strings.HasPrefix(adMarkup, `<?xml`) {
-		return smtAdTypeVideo, nil
-	} else if strings.HasPrefix(adMarkup, `{"native":`) {
-		return smtAdTypeNative, nil
 	} else {
 		return "", &errortypes.BadServerResponse{
-			Message: fmt.Sprintf("Invalid ad markup %s.", adMarkup),
+			Message: fmt.Sprintf("X-Smt-Adtype header is missing!"),
```
Comment (SyntaxNode) [inline adapters/smaato/smaato.go:274]: Nitpick: Consider changing the `!` to `.`. We keep to plain punctuation for errors.

File: adapters/smaato/banner.go (hunk 2 above)
Comment (Enigo) [reply adapters/smaato/banner.go:13]: unlikely more than 3
I believe atm we mostly supply 1

File: adapters/smaato/smaato.go (hunk 4 above)
Comment (Enigo) [reply adapters/smaato/smaato.go:61]: click url

File: adapters/smaato/smaato.go (hunk 5 above)
Comment (Enigo) [reply adapters/smaato/smaato.go:274]: fixed

File: adapters/smaato/banner.go (hunk 2 above)
Comment (Enigo) [reply adapters/smaato/banner.go:13]: I also changed it to only execute this logic when `curls` field is sent

cc @el-chuck

File: adapters/smaato/banner_test.go (hunk 3 above)
Comment (Enigo) [reply adapters/smaato/banner_test.go:42]: good catch, I added one more test case `curls=nil`
highly unlikely that we would ever return this from the server, but better have it covered

**Review by przemkaczmarek — APPROVED** [review]

File: adapters/smaato/smaato.go [hunk 6]
```diff
@@ -17,7 +17,7 @@ import (
 	"github.com/prebid/prebid-server/v2/util/timeutil"
 )
 
-const clientVersion = "prebid_server_0.7"
+const clientVersion = "prebid_server_1.1"
```
[the remark this answers is not in the document]
Comment (Enigo) [reply adapters/smaato/smaato.go:20]: Coz there are different changes, not to mix
This pr has the commit from the other pr

**Comment by bsardo** [comment]
@Enigo it looks like this PR was built off of another PR. Let's work on getting #3734 merged first and then you can merge this PR with master so that this PR only consists of the DOOH changes.

**Review by przemkaczmarek — APPROVED** [review]

**Comment by Enigo** [comment]
@przemkaczmarek @bsardo 
I've rebased this PR from the latest master, no other changes were made

**Review by przemkaczmarek — APPROVED** [review]

**Review by bsardo — APPROVED** [review]

**Pull request merged**